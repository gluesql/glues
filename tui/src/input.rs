use tui_textarea::Input as TextAreaInput;

#[derive(Clone, Debug)]
pub enum Input {
    Key(KeyEvent),
    Paste(String),
    Resize(u16, u16),
}

#[derive(Clone, Debug)]
pub struct KeyEvent {
    pub code: KeyCode,
    pub modifiers: KeyModifiers,
    pub kind: KeyEventKind,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum KeyCode {
    Char(char),
    F(u8),
    Backspace,
    Enter,
    Left,
    Right,
    Up,
    Down,
    Tab,
    Delete,
    Home,
    End,
    PageUp,
    PageDown,
    Esc,
    Null,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum KeyEventKind {
    Press,
    Release,
    Repeat,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct KeyModifiers {
    pub ctrl: bool,
    pub alt: bool,
    pub shift: bool,
}

impl KeyModifiers {
    pub const NONE: Self = Self {
        ctrl: false,
        alt: false,
        shift: false,
    };

    pub const CONTROL: Self = Self {
        ctrl: true,
        alt: false,
        shift: false,
    };

    pub fn new(ctrl: bool, alt: bool, shift: bool) -> Self {
        Self { ctrl, alt, shift }
    }
}

impl Default for KeyModifiers {
    fn default() -> Self {
        Self::NONE
    }
}

#[cfg(not(target_arch = "wasm32"))]
mod native {
    use super::*;
    use ratatui::crossterm::event as ct;

    impl From<ct::Event> for Input {
        fn from(event: ct::Event) -> Self {
            match event {
                ct::Event::Key(key) => Input::Key(key.into()),
                ct::Event::Paste(paste) => Input::Paste(paste),
                ct::Event::Resize(w, h) => Input::Resize(w, h),
                _ => Input::Resize(0, 0),
            }
        }
    }

    impl From<ct::KeyEvent> for KeyEvent {
        fn from(key: ct::KeyEvent) -> Self {
            let modifiers = KeyModifiers::new(
                key.modifiers.contains(ct::KeyModifiers::CONTROL),
                key.modifiers.contains(ct::KeyModifiers::ALT),
                key.modifiers.contains(ct::KeyModifiers::SHIFT),
            );

            let code = match key.code {
                ct::KeyCode::Char(c) => KeyCode::Char(c),
                ct::KeyCode::F(n) => KeyCode::F(n),
                ct::KeyCode::Backspace => KeyCode::Backspace,
                ct::KeyCode::Enter => KeyCode::Enter,
                ct::KeyCode::Left => KeyCode::Left,
                ct::KeyCode::Right => KeyCode::Right,
                ct::KeyCode::Up => KeyCode::Up,
                ct::KeyCode::Down => KeyCode::Down,
                ct::KeyCode::Tab => KeyCode::Tab,
                ct::KeyCode::Delete => KeyCode::Delete,
                ct::KeyCode::Home => KeyCode::Home,
                ct::KeyCode::End => KeyCode::End,
                ct::KeyCode::PageUp => KeyCode::PageUp,
                ct::KeyCode::PageDown => KeyCode::PageDown,
                ct::KeyCode::Esc => KeyCode::Esc,
                _ => KeyCode::Null,
            };

            let kind = match key.kind {
                ct::KeyEventKind::Press => KeyEventKind::Press,
                ct::KeyEventKind::Release => KeyEventKind::Release,
                ct::KeyEventKind::Repeat => KeyEventKind::Repeat,
            };

            KeyEvent {
                code,
                modifiers,
                kind,
            }
        }
    }
}

#[cfg(target_arch = "wasm32")]
mod wasm {
    use super::*;
    use ratzilla::event as rz;

    impl From<rz::KeyEvent> for KeyEvent {
        fn from(event: rz::KeyEvent) -> Self {
            let code = match event.code {
                rz::KeyCode::Char(c) => KeyCode::Char(c),
                rz::KeyCode::F(n) => KeyCode::F(n),
                rz::KeyCode::Backspace => KeyCode::Backspace,
                rz::KeyCode::Enter => KeyCode::Enter,
                rz::KeyCode::Left => KeyCode::Left,
                rz::KeyCode::Right => KeyCode::Right,
                rz::KeyCode::Up => KeyCode::Up,
                rz::KeyCode::Down => KeyCode::Down,
                rz::KeyCode::Tab => KeyCode::Tab,
                rz::KeyCode::Delete => KeyCode::Delete,
                rz::KeyCode::Home => KeyCode::Home,
                rz::KeyCode::End => KeyCode::End,
                rz::KeyCode::PageUp => KeyCode::PageUp,
                rz::KeyCode::PageDown => KeyCode::PageDown,
                rz::KeyCode::Esc => KeyCode::Esc,
                rz::KeyCode::Unidentified => KeyCode::Null,
            };

            KeyEvent {
                code,
                modifiers: KeyModifiers::new(event.ctrl, event.alt, event.shift),
                kind: KeyEventKind::Press,
            }
        }
    }
}

pub fn to_textarea_input(input: &Input) -> Option<TextAreaInput> {
    match input {
        Input::Key(key) => Some(key_to_textarea_input(key)),
        _ => None,
    }
}

pub fn key_to_textarea_input(key: &KeyEvent) -> TextAreaInput {
    use tui_textarea::Key;

    let key_code = match key.code {
        KeyCode::Char(c) => Key::Char(c),
        KeyCode::F(n) => Key::F(n),
        KeyCode::Backspace => Key::Backspace,
        KeyCode::Enter => Key::Enter,
        KeyCode::Left => Key::Left,
        KeyCode::Right => Key::Right,
        KeyCode::Up => Key::Up,
        KeyCode::Down => Key::Down,
        KeyCode::Tab => Key::Tab,
        KeyCode::Delete => Key::Delete,
        KeyCode::Home => Key::Home,
        KeyCode::End => Key::End,
        KeyCode::PageUp => Key::PageUp,
        KeyCode::PageDown => Key::PageDown,
        KeyCode::Esc => Key::Esc,
        KeyCode::Null => Key::Null,
    };

    TextAreaInput {
        key: key_code,
        ctrl: key.modifiers.ctrl,
        alt: key.modifiers.alt,
        shift: key.modifiers.shift,
    }
}
