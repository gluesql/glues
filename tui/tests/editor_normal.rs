#[macro_use]
mod tester;
use tester::Tester;

use color_eyre::Result;
use glues_tui::input::KeyCode;

#[tokio::test]
async fn quits_on_esc_then_q() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;

    t.open_first_note().await?;

    // Esc then 'q' should quit
    t.key(KeyCode::Esc).await;
    let quit = t.press('q').await;
    assert!(quit);

    Ok(())
}

#[tokio::test]
async fn quit_menu_cancel_from_normal() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;
    t.open_first_note().await?;

    t.key(KeyCode::Esc).await;
    t.draw()?;
    snap!(t, "quit_menu_open");

    t.key(KeyCode::Esc).await;
    t.draw()?;
    snap!(t, "quit_menu_closed");
    Ok(())
}

#[tokio::test]
async fn quit_menu_returns_to_entry() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;

    t.key(KeyCode::Esc).await;
    let quit = t.press('m').await;
    assert!(!quit);

    t.draw()?;
    snap!(t, "entry_after_quit_menu");
    Ok(())
}

#[tokio::test]
async fn undo_then_redo_after_insert() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;
    t.open_first_note().await?;

    // insert text
    t.press('i').await;
    t.type_str("Hello").await;
    t.key(KeyCode::Esc).await;
    t.draw()?;
    snap!(t, "text_after_insert");

    // undo (u)
    t.press('u').await;
    t.draw()?;
    snap!(t, "after_undo");

    // redo (Ctrl+r)
    t.ctrl('r').await;
    t.draw()?;
    snap!(t, "after_redo");

    Ok(())
}

#[tokio::test]
async fn yank_line_then_paste() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;
    t.open_first_note().await?;

    // yy then p
    t.press('y').await;
    t.press('y').await;
    t.press('p').await;
    t.draw()?;
    snap!(t, "after_yank_paste");

    Ok(())
}

#[tokio::test]
async fn delete_line_with_dd() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;
    t.open_first_note().await?;

    t.press('d').await;
    t.press('d').await;
    t.draw()?;
    snap!(t, "after_delete_line");

    Ok(())
}

#[tokio::test]
async fn delete_word_with_dw() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;
    t.open_first_note().await?;

    // Insert some text with multiple words
    t.press('i').await;
    t.type_str("Hello world test").await;
    t.key(KeyCode::Esc).await;

    // Move to beginning of line
    t.press('0').await;

    // Delete word with 'dw'
    t.press('d').await;
    t.press('w').await;
    t.draw()?;
    snap!(t, "after_delete_word_dw");

    Ok(())
}

#[tokio::test]
async fn delete_word_with_de() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;
    t.open_first_note().await?;

    // Insert some text with multiple words
    t.press('i').await;
    t.type_str("Hello world test").await;
    t.key(KeyCode::Esc).await;

    // Move to beginning of line
    t.press('0').await;

    // Delete word with 'de'
    t.press('d').await;
    t.press('e').await;
    t.draw()?;
    snap!(t, "after_delete_word_de");

    Ok(())
}

#[tokio::test]
async fn gateway_moves_cursor_to_top() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;
    t.open_first_note().await?;

    populate_long_note(&mut t).await?;
    t.draw()?;

    // Move to bottom so the gateway jump visibly changes the view.
    t.press('G').await;

    // First 'g' enters gateway mode.
    t.press('g').await;
    t.draw()?;
    snap!(t, "gateway_prompt");

    // Second 'g' jumps to the top line.
    t.press('g').await;
    t.draw()?;
    snap!(t, "gateway_move_top");

    Ok(())
}

#[tokio::test]
async fn scroll_commands_update_view() -> Result<()> {
    let mut t = Tester::new().await?;
    t.open_instant().await?;
    t.open_first_note().await?;

    populate_long_note(&mut t).await?;
    t.draw()?;

    // Start near the middle of the document.
    t.press('g').await;
    t.press('g').await;
    for _ in 0..20 {
        t.press('j').await;
    }

    t.press('z').await;
    t.draw()?;
    snap!(t, "scroll_prompt");

    t.press('t').await;
    t.draw()?;
    snap!(t, "scroll_jump_top");

    // Jump to bottom and use 'zb'.
    t.press('G').await;
    t.press('z').await;
    t.press('b').await;
    t.draw()?;
    snap!(t, "scroll_jump_bottom");

    // Move back toward the middle and center with 'z.'.
    for _ in 0..15 {
        t.press('k').await;
    }
    t.press('z').await;
    t.press('.').await;
    t.draw()?;
    snap!(t, "scroll_jump_center");

    Ok(())
}

async fn populate_long_note(tester: &mut Tester) -> Result<()> {
    let lines: Vec<String> = (1..=60)
        .map(|i| format!("Line {i:02} â€” the quick brown fox jumps over the lazy dog."))
        .collect();

    tester.press('g').await;
    tester.press('g').await;

    tester.press('d').await;
    tester.press('d').await;

    tester.press('i').await;
    for (idx, line) in lines.iter().enumerate() {
        tester.type_str(line).await;
        if idx + 1 < lines.len() {
            tester.key(KeyCode::Enter).await;
        }
    }
    tester.key(KeyCode::Esc).await;

    tester.key(KeyCode::Tab).await;
    tester.draw()?;
    tester.key(KeyCode::Tab).await;

    Ok(())
}
