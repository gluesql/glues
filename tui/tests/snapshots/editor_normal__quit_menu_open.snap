---
source: tui/tests/editor_normal.rs
assertion_line: 31
expression: text
snapshot_kind: text
---
 Note 'Sample Note' normal mode                                                                        [?] Show keymap 
[Browser]                                   ▐ 󱇗 Sample Note                                                             
 󰝰 Notes                                    ▐ 1 Hi :D                                                                   
   󱇗 Sample Note                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                      ┌─────────────────Notebook─────────────────┐                                      
                                      │                                          │                                      
                                      │  Leave the notebook?                     │                                      
                                      │                                          │                                      
                                      │                                          │                                      
                                      │                                          │                                      
                                      │  [m] Back to menu                        │                                      
                                      │  [q] Quit                                │                                      
                                      │  [Esc] Cancel                            │                                      
                                      │                                          │                                      
                                      └──────────────────────────────────────────┘                                      
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐                                                                           
                                            ▐ NORMAL   󰝰 Notes  󱇗 Sample Note 
